From b7423db533c98fdc0a9a72ec1255e3d4dd62babe Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Mon, 8 Jun 2026 23:36:03 -0400
Subject: [PATCH] winewayland: Don't rely on win32u keystate for
 release_all_keys.

Due to hooks. This solution is kinda a hack but whatever... shit needs to work in somehow.
---
 dlls/winewayland.drv/wayland_keyboard.c | 41 +++++++++++++------------
 dlls/winewayland.drv/waylanddrv.h       |  3 ++
 2 files changed, 25 insertions(+), 19 deletions(-)

diff --git a/dlls/winewayland.drv/wayland_keyboard.c b/dlls/winewayland.drv/wayland_keyboard.c
index e78e418c37f..133ee190840 100644
--- a/dlls/winewayland.drv/wayland_keyboard.c
+++ b/dlls/winewayland.drv/wayland_keyboard.c
@@ -837,29 +837,24 @@ static BOOL find_xkb_layout_variant(const char *name, const char **layout, const
 
 static void release_all_keys(HWND hwnd)
 {
-    BYTE state[256];
-    int vkey;
-    INPUT input = {.type = INPUT_KEYBOARD};
+    INPUT input = {0};
+    struct wayland_keyboard *keyboard = &process_wayland.keyboard;
 
-    NtUserGetAsyncKeyboardState(state);
+    input.type = INPUT_KEYBOARD;
 
-    for (vkey = 1; vkey < 256; vkey++)
+    for (UINT i = 0; i < 0x300; i++)
     {
-        /* Skip mouse buttons. */
-        if (vkey < 7 && vkey != VK_CANCEL) continue;
-        /* Skip left/right-agnostic modifier vkeys. */
-        if (vkey == VK_SHIFT || vkey == VK_CONTROL || vkey == VK_MENU) continue;
+        if (!keyboard->keystate[i]) continue;
 
-        if (state[vkey] & 0x80)
-        {
-            UINT scan = NtUserMapVirtualKeyEx(vkey, MAPVK_VK_TO_VSC_EX,
-                                              keyboard_hkl);
-            input.ki.wVk = vkey;
-            input.ki.wScan = scan & 0xff;
-            input.ki.dwFlags = KEYEVENTF_KEYUP;
-            if (scan & ~0xff) input.ki.dwFlags |= KEYEVENTF_EXTENDEDKEY;
-            NtUserSendHardwareInput(hwnd, 0, &input, 0);
-        }
+        input.ki.dwFlags = KEYEVENTF_SCANCODE | KEYEVENTF_KEYUP;
+        input.ki.wScan = (i & 0x300) ? i + 0xdf00 : i;
+
+        if (i & 0x300) input.ki.dwFlags |= KEYEVENTF_EXTENDEDKEY;
+        if (i == key2scan(KEY_RIGHTSHIFT) || i == key2scan(KEY_NUMLOCK))
+            input.ki.dwFlags |= KEYEVENTF_EXTENDEDKEY;
+
+        NtUserSendHardwareInput(hwnd, 0, &input, 0);
+        keyboard->keystate[i] = 0;
     }
 }
 
@@ -1071,6 +1066,12 @@ static void keyboard_handle_leave(void *private, struct wl_keyboard *wl_keyboard
     }
 }
 
+static inline void update_keystate(uint32_t key, uint32_t state)
+{
+    struct wayland_keyboard *keyboard = &process_wayland.keyboard;
+    keyboard->keystate[key2scan(key)] = (state == WL_KEYBOARD_KEY_STATE_PRESSED);
+}
+
 static void send_right_control(HWND hwnd, uint32_t state)
 {
     INPUT input = {0};
@@ -1079,6 +1080,7 @@ static void send_right_control(HWND hwnd, uint32_t state)
     input.ki.dwFlags = KEYEVENTF_SCANCODE | KEYEVENTF_EXTENDEDKEY;
     if (state == WL_KEYBOARD_KEY_STATE_RELEASED) input.ki.dwFlags |= KEYEVENTF_KEYUP;
     NtUserSendHardwareInput(hwnd, 0, &input, 0);
+    update_keystate(KEY_RIGHTCTRL, state);
 }
 
 static void set_async_key_state(const BYTE state[256])
@@ -1166,6 +1168,7 @@ static void keyboard_handle_key(void *data, struct wl_keyboard *wl_keyboard,
 
     if (state == WL_KEYBOARD_KEY_STATE_RELEASED) input.ki.dwFlags |= KEYEVENTF_KEYUP;
     NtUserSendHardwareInput(hwnd, 0, &input, 0);
+    update_keystate(key, state);
 }
 
 static void keyboard_handle_modifiers(void *data, struct wl_keyboard *wl_keyboard,
diff --git a/dlls/winewayland.drv/waylanddrv.h b/dlls/winewayland.drv/waylanddrv.h
index f8ff7615dfa..ee99b205d0f 100644
--- a/dlls/winewayland.drv/waylanddrv.h
+++ b/dlls/winewayland.drv/waylanddrv.h
@@ -113,6 +113,9 @@ struct wayland_keyboard
     struct xkb_state *xkb_state;
     HWND focused_hwnd;
     pthread_mutex_t mutex;
+    /* scancode -> keystate mapping. The keyboard state is the same across every instance of a seat.
+     * We bind to the first seat each time, and this first bind seat should be the same every time. */
+    unsigned char keystate[0x300];
 };
 
 struct wayland_cursor
-- 
2.56.0

